src/libpmemfile-posix/inode.c

```diff
@@ -0,0 +1,841 @@
+		}
+	}
+	os_rwlock_unlock(&map->rwlock);
+
+	if (is_new) {
+		rwlock_tx_wlock(&map->rwlock);
+		tx = 1;
+	} else
+		os_rwlock_wrlock(&map->rwlock);
+
+	/* recalculate slot, someone could rebuild the hash map */
+	idx = inode_hash(map, inode) % map->sz;
+
+	/* check again */
+	b = &map->buckets[idx];
+	unsigned empty_slot = UINT32_MAX;
+	for (unsigned j = 0; j < BUCKET_SIZE; ++j) {
+		if (TOID_EQUALS(b->arr[j].pinode, inode)) {
+			vinode = b->arr[j].vinode;
+			goto end;
+		}
+		if (empty_slot == UINT32_MAX && b->arr[j].pinode.oid.off == 0)
+			empty_slot = j;
+	}
+
+	int tries = 0;
+	while (empty_slot == UINT32_MAX) {
+		size_t new_sz = map->sz;
+
+		do {
+			if (map->inodes > 2 * new_sz || tries == 2) {
+				new_sz *= 2;
+				tries = 0;
+			} else {
+				inode_map_rand_params(map);
+				tries++;
+			}
+		} while (!inode_map_rebuild(map, new_sz));
+
+		idx = inode_hash(map, inode) % map->sz;
+		b = &map->buckets[idx];
+
+		for (unsigned j = 0; j < BUCKET_SIZE; ++j) {
+			if (b->arr[j].pinode.oid.off == 0) {
+				empty_slot = j;
+				break;
+			}
+		}
+	}
+
+	vinode = calloc(1, sizeof(*vinode));
+	if (!vinode) {
+		error = errno;
+		goto end;
+	}
+
+	os_rwlock_init(&vinode->rwlock);
+	vinode->tinode = inode;
+	vinode->inode = D_RW(inode);
+	if (inode_is_dir(vinode->inode) && parent) {
+		vinode->parent = vinode_ref(pfp, parent);
+
+		if (parent_refed)
+			*parent_refed = true;
+	}
+
+	if (parent && name && namelen)
+		vinode_set_debug_path_locked(pfp, parent, vinode, name,
+				namelen);
+
+	b->arr[empty_slot].pinode = inode;
+	b->arr[empty_slot].vinode = vinode;
+	map->inodes++;
+
+	if (is_new)
+		cb_push_front(TX_STAGE_ONABORT,
+			(cb_basic)vinode_unregister_locked,
+			vinode);
+
+end:
+	__sync_fetch_and_add(&vinode->ref, 1);
+	if (is_new && tx)
+		rwlock_tx_unlock_on_commit(&map->rwlock);
+	else
+		os_rwlock_unlock(&map->rwlock);
+
+	if (error)
+		errno = error;
+
+	return vinode;
+}
+
+/*
+ * inode_ref_new -- increases inode reference counter
+ *
+ * Assumes inode was allocated in the same transaction.
+ * Return volatile inode.
+ */
+struct pmemfile_vinode *
+inode_ref_new(PMEMfilepool *pfp,
+		TOID(struct pmemfile_inode) inode,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen)
+{
+	return _inode_get(pfp, inode, true, parent, parent_refed, name,
+			namelen);
+}
+
+/*
+ * inode_ref -- increases inode reference counter
+ *
+ * Assumes inode was not allocated in the same transaction.
+ * Return volatile inode.
+ */
+struct pmemfile_vinode *
+inode_ref(PMEMfilepool *pfp,
+		TOID(struct pmemfile_inode) inode,
+		struct pmemfile_vinode *parent,
+		volatile bool *parent_refed,
+		const char *name,
+		size_t namelen)
+{
+	return _inode_get(pfp, inode, false, parent, parent_refed, name,
+			namelen);
+}
+
+/*
+ * vinode_tx_unref -- decreases inode reference counter
+ *
+ * Must be called in a transaction.
+ */
+static bool
+vinode_tx_unref(PMEMfilepool *pfp, struct pmemfile_vinode *vinode)
```

Comment:
This function is only called from vinode_unref. The actual 'transactional' part of this code is in inode_array_unregister.
The comments are misleading.
I would consider folding those two functions together, the distinction seems artificial.